Performance best practice update with davidfowl samples

Rick-Anderson opened the pull request:
Fixes #14605
[Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/performance/performance-best-practices?view=aspnetcore-3.1&branch=pr-en-us-14734#avoid-using-synchronous-readwrite-overloads-on-httprequestbody-and-httpresponsebody)

davidfowl commented:
@Rick-Anderson where's the review url?

Rick-Anderson commented:
@davidfowl  This is WIP. I forgot to do the draft when I created the PR. I'm just getting started, not ready for review.

[Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/performance/performance-best-practices?view=aspnetcore-3.1&branch=pr-en-us-14734)

davidfowl commented on aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncBadVoidController.cs:
```diff
@@ -15,8 +15,6 @@ public async void Get()
         {
             await Task.Delay(1000);
 
-            // THIS will crash the process because it's writing after the response has 
```
Why did you remove this comment?

Rick-Anderson replied:
> Why did you remove this comment?


See my  notes below. We generally move all comments to the doc so they can be MT.

Rick-Anderson replied:
Let me know which ones you want back and I can bend the rules and add a few.

Rick-Anderson replied:
Edited and added back :)

Rick-Anderson commented:
@davidfowl  ready for review at [Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/performance/performance-best-practices?view=aspnetcore-3.1&branch=pr-en-us-14734#avoid-using-synchronous-readwrite-overloads-on-httprequestbody-and-httpresponsebody)

Please search for `<!-- TODO Review`  I have questions on those two sections.

* Moved all code comments out of code and into the document. Code comments are not MT (Machine Translated).
* Changed code to get it to compile. For example you use `app.Use(async (next, context) =>` and I changed that to `app.Use(async (context, next) =>`
* Shortened sentences so they would MT. Even native readers can comprehend more with shorter sentences.

davidfowl commented on aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncFirstController.cs:
```diff
@@ -0,0 +1,141 @@
+            try
+            {
+                _logger.LogInformation("Starting search query from {path}.", 
+                                                             HttpContext.Request.Path);
```
The formatting here is weird. Can we fix it?

davidfowl commented on aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncFirstController.cs:
```diff
@@ -0,0 +1,141 @@
+        public readonly SearchService _searchService;
+
+        public AsynchBadSearchController(ILogger<AsynchBadSearchController> logger,
+                                                        SearchService searchService)
```
The formatting here is weird too. Can we left align the parameters?

davidfowl commented on aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncFirstController.cs:
```diff
@@ -0,0 +1,141 @@
+        {
+            string path = HttpContext.Request.Path;
+            var query1 = SearchAsync(SearchEngine.Google, query,
+                path);
```
Same here?

davidfowl commented on aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncFirstController.cs:
```diff
@@ -0,0 +1,141 @@
+        public static SearchEngine Bing { get; internal set; }
+        public static SearchEngine Google { get; internal set; }
+        public static SearchEngine DuckDuckGo { get; internal set; }
+
```
nit: Remove extra space

davidfowl commented on aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncFirstController.cs:
```diff
@@ -0,0 +1,141 @@
+            _logger = logger;
+            _searchService = searchService;
+        }
+
```
nit: Remove extra space

davidfowl commented on aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncFirstController.cs:
```diff
@@ -0,0 +1,141 @@
+            throw new NotImplementedException();
+        }
+    }
+
```
nit: Remove extra space

davidfowl commented on aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/FireAndForgetSecondController.cs:
```diff
@@ -0,0 +1,77 @@
+
+        internal Task SaveChangesAsync()
+        {
+            throw new NotImplementedException();
```
Make this return Task.CompletedTask

davidfowl commented on aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/MyFirstController.cs:
```diff
@@ -0,0 +1,41 @@
+        [HttpGet("/pokemon")]
+        public async Task<ActionResult<PokemonData>> Get()
+        {
+            var json = await new StreamReader(Request.Body).ReadToEndAsync();
```
Since we have an async JSON serializer, lets change this to:

```C#
return await JsonSerializer.DeserializeAsync<PokemonData>(Request.Body);
```

davidfowl replied:
Remove the StreamReader logic.

davidfowl commented on aspnetcore/performance/performance-best-practices/samples/3.0/Program.cs:
```diff
@@ -0,0 +1,31 @@
+                 logging.ClearProviders();
+                 logging.AddConsole();
+             })
+                .ConfigureWebHostDefaults(webBuilder =>
```
nit: Fix formatting here to align the code properly

davidfowl commented on aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncFirstController.cs:
```diff
@@ -0,0 +1,141 @@
+        private readonly ILogger<AsynchBadSearchController> _logger;
+        public readonly SearchService _searchService;
+
+        public AsynchBadSearchController(ILogger<AsynchBadSearchController> logger,
```
Remove the h from `Asynch`

davidfowl commented on aspnetcore/performance/performance-best-practices/samples/3.0/Startup.cs:
```diff
@@ -0,0 +1,110 @@
+        }
+    }
+
+    public class SearchService : IDisposable {
```
What's this for?

Rick-Anderson replied:
> What's this for?
Auto code generation by VS to get the code to compile. It offered a fix and I clicked "OK". Please suggest better code.

davidfowl replied:
Remove it all and just implement the interface 😄 .

Rick-Anderson replied:
`public class SearchService : IDisposable {`

> What's this for?

VS generated code so your snippet would compile.

davidfowl commented on aspnetcore/performance/performance-best-practices.md:
```diff
@@ -132,19 +131,200 @@ Recommendations:
+
+[!code-csharp[](performance-best-practices/samples/3.0/MyType.cs?name=snippet2)]
+
+## Do not access HttpContext from multiple threads
```
The highlights are messed up in this section.

![image](https://user-images.githubusercontent.com/95136/66107853-1e5fb080-e576-11e9-913d-ba58f6d7459e.png)

davidfowl replied:
it's still busted https://review.docs.microsoft.com/en-us/aspnet/core/performance/performance-best-practices?view=aspnetcore-3.1&branch=pr-en-us-14734#do-not-access-httpcontext-from-multiple-threads

Rick-Anderson replied:
It's correct on edge/FF and Chrome at 80%. The build team (who writes the HTML generator for all of doc.ms.com) broke it 6 months ago and have only managed to mostly fix it.

Edge:
![image](https://user-images.githubusercontent.com/3605364/66226387-62d27580-e676-11e9-92cf-496c45a33743.png)

FF:
![image](https://user-images.githubusercontent.com/3605364/66226410-71b92800-e676-11e9-922f-d870c6989f47.png)

Chrome at 80%:

![image](https://user-images.githubusercontent.com/3605364/66226460-8bf30600-e676-11e9-8909-de32958d8f43.png)

Rick-Anderson replied:
@davidfowl  see #14936
I asked for the problem to be  prioritized on the internal bug.

davidfowl commented on aspnetcore/performance/performance-best-practices.md:
```diff
@@ -132,19 +131,200 @@ Recommendations:
+
+[!code-csharp[](performance-best-practices/samples/3.0/Controllers/MyFirstController.cs?name=snippet2)]
+
+The preceding code asynchronously reads the entire HTTP request body into memory.
```
If we remove the ReadToEndAsync and show the async JSON reading directly then it's fine to remove this

halter73 commented on aspnetcore/performance/performance-best-practices/samples/3.0/Startup22.cs:
```diff
@@ -0,0 +1,54 @@
+            #region snippet2
+            app.Use(async (context, next) =>
+            {
+                await context.Response.WriteAsync("Hello ");
```
Remove this line. With this here, it makes no sense why we're checking `context.Response.HasStarted` since it would almost\* always be true. If only `next()` is called, it remains ambiguous whether or not the response started.

\* `context.Response.HasStarted` could still be false if the response body Stream was replaced before attempting to write to it, but since it's not mentioned anywhere that may have happened, I'd just remove the attempted write.

halter73 commented on aspnetcore/performance/performance-best-practices/samples/3.0/Startup22.cs:
```diff
@@ -0,0 +1,54 @@
+            #region snippet1
+            app.Use(async (context, next) =>
+            {
+                await context.Response.WriteAsync("Hello ");
```
I'd remove this line too. Even though this is the "bad" example, it's supposed to be bad because we're not checking `context.Response.HasStarted`, not just because we wrote to the response body and then tried changing the response header (which is also bad ofc).

davidfowl commented:
@Rick-Anderson can we rename the section to:

Performance and Reliability?

Tratcher commented on aspnetcore/performance/performance-best-practices.md:
```diff
@@ -132,19 +131,200 @@ Recommendations:
+
+When using a serializer/de-serializer that only supports synchronous reads and writes:
+
+* Choose to buffer the data into memory before passing data into the serializer/de-serializer.
```
```suggestion
* Buffer the data into memory asynchronously before passing it into the serializer/de-serializer.
```

Tratcher commented on aspnetcore/performance/performance-best-practices.md:
```diff
@@ -132,19 +131,200 @@ Recommendations:
+When using a serializer/de-serializer that only supports synchronous reads and writes:
+
+* Choose to buffer the data into memory before passing data into the serializer/de-serializer.
+* [JSON.NET](https://www.newtonsoft.com/json/help/html/Introduction.htm) is only synchronous.
```
This bullet point is in the wrong place. It's an example of a sync serializer, not an action to take when using one.

Tratcher commented on aspnetcore/performance/performance-best-practices.md:
```diff
@@ -132,19 +131,200 @@ Recommendations:
+* May result in quickly running out of space in the LOH.
+* May cause performance issues for the app because of full GCs running.
+
+## Use buffered and synchronous reads and writes as an alternative to asynchronous reading and writing
```
Working with a synchronous data processing API

Tratcher commented on aspnetcore/performance/performance-best-practices.md:
```diff
@@ -132,19 +131,200 @@ Recommendations:
+
+## Use buffered and synchronous reads and writes as an alternative to asynchronous reading and writing
+
+When using a serializer/de-serializer that only supports synchronous reads and writes:
```
like JSON.NET...

Tratcher commented on aspnetcore/performance/performance-best-practices.md:
```diff
@@ -132,19 +131,200 @@ Recommendations:
+
+[!code-csharp[](performance-best-practices/samples/3.0/Controllers/FireAndForgetSecondController.cs?name=snippet2&highlight=9-16)]
+
+## Avoid adding headers after the HttpResponse has started
```
```suggestion
## Do not modify the status code or headers after the response body has started
```

Tratcher replied:
Related: Do not call `next()` if you have already started writing to the response body. Components only expect to be called if it's possible for them to handle/manipulate the response.

Tratcher commented on aspnetcore/performance/performance-best-practices.md:
```diff
@@ -132,19 +131,200 @@ Recommendations:
+
+[!code-csharp[](performance-best-practices/samples/3.0/Startup22.cs?name=snippet1)]
+
+In the preceding code, `context.Response.Headers["test"] = "test value";` if `next()` has written to the response.
```
```suggestion
In the preceding code, `context.Response.Headers["test"] = "test value";` will throw an exception if `next()` has written to the response.
```

Tratcher commented on aspnetcore/performance/performance-best-practices.md:
```diff
@@ -132,19 +131,200 @@ Recommendations:
+
+In the preceding code, `context.Response.Headers["test"] = "test value";` if `next()` has written to the response.
+
+**Do this:** The following example checks if the HTTP response has started before writing to the body.
```
```suggestion
**Do this:** The following example checks if the HTTP response has started before modifying the headers.
```

Tratcher commented on aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/MySecondController.cs:
```diff
@@ -0,0 +1,58 @@
+        [HttpPost("/form-body")]
+        public IActionResult Post()
+        {
+            var form =  HttpContext.Request.Form;
```
This is misleading to show in a controller, MVC usually reads the form for you async during model binding.

Rick-Anderson replied:
@Tratcher  this comes from https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AspNetCoreGuidance.md#prefer-using-httprequestreadasformasync-over-httprequestform
I'm not allowed to change code so you'll have to provide new code.

davidfowl replied:
We can change it to middleware but the point is to show both

davidfowl commented:
@Rick-Anderson the highlighting in the "Do not access HttpContext from multiple threads" is still busted but I think after it's fixed, we should merge this.
